infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/fsm/LogReplicationFSM.java

```diff
@@ -39,64 +40,84 @@
+ *             |                             |   |  |      |                               |    CANCEL
+ *             |                             |   |  |      |                               |   +------+
+ *             |                             |   |  |      |                               |   |      |
+ *             |                             |   |  |      |                               |   |      |
+ * +-----------v-----------+   REPLICATION   |   |  |      |   REPLICATION     +-----------v---+------v--+
+ * |   IN_LOG_ENTRY_SYNC   +------STOP-------+   |  |      +------STOP---------+    IN_SNAPSHOT_SYNC     |
+ * +----+-----^---+------^-+                     |  |                          +-^-----+---^-----+-----^-+
+ *      |     |   |      |                       |  |                            |     |   |     |     |
+ *      |     |   |      |                       |  |                            |     |   |     |     |
+ *      |     |   +------+           REPLICATION |  | SNAPSHOT_TRANSFER          +-----+   |     |     |
+ *      |     |   LOG_ENTRY              STOP    |  |      COMPLETE              SNAPSHOT  |     |     |
+ *      |     | SYNC_CONTINUE                    |  |                              SYNC    |     |     |
+ *      |     |                                  |  |                            CONTINUE  |     |     |
+ *      |     |                                  |  |                                      |     |     |
+ *      |     |                                  |  |                                      |     |     |
+ *      |     |                                  |  |                                      |     |     |
+ *      |     |                         +--------+--v-------------+  SNAPSHOT_SYNC_REQUEST |     |     |
+ *      |     +-----SNAPSHOT_APPLY------+   WAIT_SNAPSHOT_APPLY   +------------------------+     |     |
+ *      |              COMPLETE         +-^----+---^--------------+       SYNC_CANCEL            |     |
+ *      |                                 |    |   |                                             |     |
+ *      |                                 |    |   +-----------SNAPSHOT_TRANSFER_COMPLETE--------+     |
+ *      |                                 +----+                                                       |
+ *      |                             SNAPSHOT_APPLY                                                   |
+ *      |                               IN_PROGRESS                                                    |
+ *      |                                                                                              |
+ *      +----------------------------------------SYNC_CANCEL-------------------------------------------+
+ *                                          SNAPSHOT_SYNC_REQUEST
+ *
+ *
+ *
+ *
+ *     +-------------+
+ *     |   STOPPED   <-----REPLICATION-----  ALL_STATES
+ *     +-------------+       SHUTDOWN
```

Comment:
How  IN_SNAPSHOT_SYNC get a SNAPSHOT_REQUEST or SNAPSHOT_CANCEL and go to WAIT_SNAPSHOT_APPLY ?

Reply:
I'm not sure I understand the question. The only event that takes us from IN_SNAPSHOT_SYNC to WAIT_SNAPSHOT_APPLY is SNAPSHOT_TRANSFER_COMPLETE, which is achieved when the ACK of type snapshot_transfer_complete is received.

Reply:
The diagram shows that is  link from IN_SNAPSHOT_SYNC to WAIT_SNAPSHOT_APPLY with mark of  SNAPSHOT_REQUEST or SNAPSHOT_CANCEL.

Reply:
The arrow is in the other direction from WAIT_SNAPSHOT_APPLY to IN_SNAPSHOT_SYNC, in case a new request comes in...